usaspending_api/search/v2/views/spending_by_award.py

```diff
@@ -0,0 +1,259 @@
+from usaspending_api.awards.v2.filters.sub_award import subaward_filter
+from usaspending_api.awards.v2.lookups.lookups import (
+    assistance_type_mapping,
+    contract_subaward_mapping,
+    contract_type_mapping,
+    direct_payment_type_mapping,
+    grant_subaward_mapping,
+    grant_type_mapping,
+    idv_type_mapping,
+    loan_type_mapping,
+    non_loan_assistance_type_mapping,
+    other_type_mapping,
+    procurement_type_mapping,
+)
+from usaspending_api.awards.v2.lookups.matview_lookups import (
+    award_contracts_mapping,
+    award_idv_mapping,
+    loan_award_mapping,
+    non_loan_assistance_award_mapping,
+)
+from usaspending_api.common.api_versioning import api_transformations, API_TRANSFORM_FUNCTIONS
+from usaspending_api.common.cache_decorator import cache_response
+from usaspending_api.common.exceptions import InvalidParameterException, UnprocessableEntityException
+from usaspending_api.common.helpers.orm_helpers import obtain_view_from_award_group
+from usaspending_api.common.validator.award_filter import AWARD_FILTER
+from usaspending_api.common.validator.pagination import PAGINATION
+from usaspending_api.common.validator.tinyshield import TinyShield
+
+
+@api_transformations(api_version=settings.API_VERSION, function_list=API_TRANSFORM_FUNCTIONS)
+class SpendingByAwardVisualizationViewSet(APIView):
+    """
+    This route takes award filters and fields, and returns the fields of the filtered awards.
+    """
+
+    endpoint_doc = "usaspending_api/api_docs/api_documentation/advanced_award_search/spending_by_award.md"
+
+    @cache_response()
+    def post(self, request):
+        """Return all awards matching the provided filters and limits"""
+        models = [
+            {"name": "fields", "key": "fields", "type": "array", "array_type": "text", "text_type": "search", "min": 1},
+            {"name": "subawards", "key": "subawards", "type": "boolean", "default": False},
+        ]
+        models.extend(copy.deepcopy(AWARD_FILTER))
+        models.extend(copy.deepcopy(PAGINATION))
+        for m in models:
+            if m["name"] in ("award_type_codes", "fields"):
+                m["optional"] = False
+
+        json_request = TinyShield(models).block(request.data)
+        self.fields = json_request["fields"]
+        filters = json_request.get("filters", {})
+        subawards = json_request["subawards"]
+        self.order = json_request["order"]
+        self.limit = json_request["limit"]
+        self.page = json_request["page"]
+        self.lower_bound = (self.page - 1) * self.limit
+        self.upper_bound = self.page * self.limit + 1
+
+        if "no intersection" in filters["award_type_codes"]:
+            # "Special case": there will never be results when the website provides this value
+            return Response(
+                {"limit": self.limit, "results": [], "page_metadata": {"page": self.page, "hasNext": False}}
```

Comment:
These key names are getting used a lot, as is the general json structure. I would suggest a factory method to return this dict given values, and using that rather than constants or duplicating stuff.

Reply:
Sounds like a lot of work for only a little benefit.

Our API is terribly inconsistent. Which makes any attempt to standardize existing API input or responses difficult.